### PR TITLE
Allow to use system tzdata instead of embedded

### PR DESCRIPTION
**Pull request by zvonand**
Allow to use system tzdata instead of ClickHouse built-in. Solution for #15361.  
Add a new global server setting `prefer_system_tzdata` allowing to choose whether embedded tzdata or system tzdata will be used.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new global server setting `prefer_system_tzdata` allowing to choose whether embedded tzdata or system tzdata will be used.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
This is very bad idea.

**Comment by UnamedRus** [comment]
> They typically use a fixed and outdated version. 

It's not even outdated.

23.3 was still LTS, but tzdata wasn't backported to it: https://github.com/ClickHouse/ClickHouse/pull/60826

**Comment by Algunenano** [comment]
I can't say much about whether this is useful or not for users and what's the security and usability concerns of allowing it (@alexey-milovidov can).

What I can say is that the proposed code is absolutely not ok. There is no way we could accept passing options in the process via getenv / setenv. That being said, there is no point on discussing code until it's clearer if we want to have this option or not,

**Comment by zvonand** [comment]
> There is no way we could accept passing options in the process via getenv / seten

Timezone (from config.xml) is already being passed from config to cctz via setenv/getenv

**Comment by filimonov** [comment]
Setting the env by the process it quite controversial indeed, i guess @zvonand just followed the same pattern which already was used before: https://github.com/ClickHouse/ClickHouse/commit/d5f8e0347ffd37b048d964a3982e12f9356b4fce - passing something to singleton objects often ends up with some dirty hacks). 

Reading the env (set outside) i see the clear usecase for clickhouse-local & clickhouse-client (it can be much handier than creating config files).

File: src/Common/DateLUTImpl.cpp [hunk 1]
```diff
@@ -284,6 +285,13 @@ namespace cctz_extension
             const std::string & name,
             const std::function<std::unique_ptr<cctz::ZoneInfoSource>(const std::string & name)> & fallback)
         {
+            const char * prefer_system_tzdata_env = std::getenv("PREFER_SYSTEM_TZDATA"); // NOLINT(concurrency-mt-unsafe)
+            if (prefer_system_tzdata_env && 0 != strcmp(prefer_system_tzdata_env, "0"))
```
Comment (filimonov) [inline src/Common/DateLUTImpl.cpp:289]: PREFER_SYSTEM_TZDATA=false?

File: src/Daemon/BaseDaemon.cpp [hunk 2]
```diff
@@ -846,6 +846,13 @@ void BaseDaemon::initialize(Application & self)
         DateLUT::setDefaultTimezone(config_timezone);
     }
 
+    if (config().has("prefer_system_tzdata"))
+    {
+        const bool prefer_system_tzdata = config().getBool("prefer_system_tzdata");
```
Comment (filimonov) [inline src/Daemon/BaseDaemon.cpp:851]: would be nice to warn about that. But it's not the best place in the code to do some logging... 

Maybe we can read those envs & do some debug prints about TZ & prefer_system_tzdata somewhere in Server.cpp?

File: src/Common/DateLUTImpl.cpp (hunk 1 above)
Comment (zvonand) [reply src/Common/DateLUTImpl.cpp:289]: what do you mean?

Comment (filimonov) [inline src/Common/DateLUTImpl.cpp:289]:
```suggestion
            if (prefer_system_tzdata_env && 
                0 != strcmp(prefer_system_tzdata_env, "0") && 
                0 != strcasecmp(prefer_system_tzdata_env, "false"))
```

I.e. should we ready that somebody will set it to false explicitly?

**Comment by zvonand** [comment]
> What I can say is that the proposed code is absolutely not ok. There is no way we could accept passing options in the process via getenv / setenv. 

Now it doesn't use env anymore, I hope this looks better

**Comment by zvonand** [comment]
> This is a very bad idea.

In general, this doesn't change anything. If there is something wrong with system tzdata, CH would still use the embedded version. I can't think of any new potentially harmful use cases.

File: src/Daemon/BaseDaemon.cpp [hunk 3]
```diff
@@ -837,7 +837,12 @@ void BaseDaemon::initialize(Application & self)
         }
     }
 
-    /// This must be done before any usage of DateLUT. In particular, before any logging.
+    /// These two parameters must be processed before any usage of DateLUT. In particular, before any logging.
```
Comment (ilejn) [inline src/Daemon/BaseDaemon.cpp:840]: "Two parameters"? Which ones?

File: src/Daemon/BaseDaemon.cpp [hunk 4]
```diff
@@ -837,7 +837,12 @@ void BaseDaemon::initialize(Application & self)
...
+    if (config().has("prefer_system_tzdata"))
```
Comment (ilejn) [inline src/Daemon/BaseDaemon.cpp:841]: Before this PR zone_info_source_factory is set regardless if BaseDaemon::initialize was called or not.

So the concern is 
1. do we have application (clickhouse-local ?) that bypasses BaseDaemon::initialize ?
2. if yes, is it a problem ?

Comment (zvonand) [reply src/Daemon/BaseDaemon.cpp:841]: Looks like ch-local does not respect this parameter indeed.
Idk if this is a major problem -- we already have config parameters that are not respected by ch-local. E.g. take a look at the `timezone` parameter that is processed below

Comment (zvonand) [reply src/Daemon/BaseDaemon.cpp:841]: I never really used CH-local, but for me it also looks like it doesn't respect logging settings that are specified in `config.xml`

Comment (ilejn) [reply src/Daemon/BaseDaemon.cpp:842]: ch-local definitely operates with dates in general.
How it would work if DateLUT::setTZDataSource is not called?

Comment (zvonand) [reply src/Daemon/BaseDaemon.cpp:842]: Actually, this is an issue, you're right.
It will use system tzdata then.

Comment (zvonand) [reply src/Daemon/BaseDaemon.cpp]: Fixed now

**Comment by zvonand** [comment]
> This is a very bad idea.

@alexey-milovidov Why you think so? The implementation itself may be changed, but what is wrong with the idea in general? In case this setting isn't modified (the default case) everything stays as is. Of course, one shall not modify this setting without knowing what is it doing, but I guess this can be said about every setting.  
(as @Algunenano said: `I can't say much about whether this is useful or not for users and what's the security and usability concerns of allowing it (@alexey-milovidov can).`)

**Comment by alexey-milovidov** [comment]
@zvonand A test for LDAP has failed: https://s3.amazonaws.com/clickhouse-test-reports/65562/57750540c4cbad4b3ea6719c14eb4611f67663f8/integration_tests__aarch64__[4_6].html

**Comment by zvonand** [comment]
@alexey-milovidov why do you think it is a bad idea?

It changes nothing for a user unless he changes this behavior, and even after that, the worse thing that could happen is CH would use built-in tzdata (as fallback)

**Comment by UnamedRus** [comment]
> can be even different across servers with the same ClickHouse version. In the case of builtin tzdata, it is fully controlled by the ClickHouse version

Problem here, that ClickHouse already take timezone in which it belongs from linux as granted. 
So, if someone changes time zone of instance, it leads to this problem with partition by https://github.com/ClickHouse/ClickHouse/issues/43399
And people step in that silent trap quite often (noone is expecting that change of timezone can break your data), we can only imagine how much such cases are not acknowledged.

**Comment by zvonand** [comment]
>  if it is very small, and not be used by reasonable customers.

Well, it is turned off by default, and it is a standalone thing, so no one will be affected implicitly.

One shall not change a setting if he doesn’t know how to work with it, this can be said about many other settings as well

**Comment by filimonov** [comment]
> Let's suppose you created a partition by `toDate(event_time)` in the server's timezone or in a certain timezone. Then, after upgrading tzdata, the timezone changes, and the data no longer belongs to the correct partition.

Normally the changes of tzdata impacts the dates in FUTURE.

I.e. government of county `X` make a decision to change the timezone rules for country XXX staring from some future date `D`.
They adjust the local laws and publish that (usually months before `D`). After that tzdata maintainers create a new version of tzdata and publish that (still far before `D`). After that responsible & reasonable admins (especially those in country `X`) just upgrade tzdata package on the servers their maintain to prevent problems after the date `X`. So when the date `X` finally comes all the servers are ready for that. 

That pattern does not work with ClickHouse. And as i mentioned in the previous comment upgrade of ClickHouse is not always an option available. 

The issue you describe may happen if you do upgrade of tzdata AFTER the date `X`. When tzdata is embedded the risks that it will be upgraded later than that date is much higher (planning & deploying of clickhouse upgrade is much 'slower' process). 
That issue is even more common when the admins change the timezone of the server or move the date (like backup & restore) between the servers with different timezones. (i can imagine some options to make those scenarios less problematic / more robust - but they are rather quite complex). 

> In the case of system tzdata it happens in an unpredictable moment in time, and can be even different across servers with the same ClickHouse version.

The only reasonable rule here - is 'never change the tzrules for the past'. Changes for future date are ok (of course if the database stores timestamp for the future, which is uncommon - then it can still  impact the ClickHouse user)

**Comment by alexey-milovidov** [comment]
@zvonand, a test for LDAP has failed in the master branch: https://s3.amazonaws.com/clickhouse-test-reports/0/c945209427e759dccfaff5a6af11f2e18c100390/integration_tests__aarch64__[4_6].html

**Comment by zvonand** [comment]
@alexey-milovidov Yes, I already seen it and thinking about it, even without 3 reminders in 3 unrelated PRs.  
By the way, there is a standalone issue for this test failure: #65762

**Comment by alexey-milovidov** [comment]
@zvonand, this test has failed: https://s3.amazonaws.com/clickhouse-test-reports/0/0aa30b10d5b4a33c7b088bfcaeaeff2620ef73ec/stateless_tests__azure__asan__%5B3_3%5D.html

**Comment by zvonand** [comment]
> @zvonand, this test has failed: https://s3.amazonaws.com/clickhouse-test-reports/0/0aa30b10d5b4a33c7b088bfcaeaeff2620ef73ec/stateless_tests__azure__asan__%5B3_3%5D.html

Is this test related to me? I just noticed the failure, did not have an idea on how to fix it and thus opened the issue

**Comment by zvonand** [comment]
@alexey-milovidov https://github.com/ClickHouse/ClickHouse/issues/67842#issuecomment-2284460087

**Comment by alexey-milovidov** [comment]
@zvonand, maybe you can help fixing this test?

**Comment by zvonand** [comment]
> maybe you can help fixing this test?

I'd be glad to, but I have no idea about how to stabilize it yet

**Comment by alexey-milovidov** [comment]
@zvonand Maybe it's worth looking to this test about "zero-copy replication" (experimental feature): #68261

**Comment by alexey-milovidov** [comment]
@zvonand, a test for DNS cache has failed: https://s3.amazonaws.com/clickhouse-test-reports/0/a129d07eb58caa153f4ddae4ef60c033f94e5965/integration_tests__tsan__[3_6].html

**Comment by ilejn** [comment]
@alexey-milovidov ,
consider https://github.com/ClickHouse/ClickHouse/pull/69121

**Comment by alexey-milovidov** [comment]
@zvonand a test for LDAP has failed: https://s3.amazonaws.com/clickhouse-test-reports/70660/75916285535e3855b62ed481661613e3eef2fbe0/integration_tests__aarch64__[5_6].html

**Comment by alexey-milovidov** [comment]
This is ok, but it provides a lot of variability in configuration - reasoning about ClickHouse behavior will be harder with this change.

**Comment by zvonand** [comment]
> reasoning about ClickHouse behavior will be harder

WDYM? If you do not change this setting -- CH behavior is same as usual. If you change it -- you know what you are doing, by default it is turned off

**Comment by alexey-milovidov** [comment]
ClickHouse is an open-source DBMS used by а countless number of companies.

If they use obscure settings or a modified build, I will have more trouble when they ask me for help.

**Pull request closed without merging**